### PR TITLE
HPCC-18762 Avoid buffer overrun writing string to ptree IXMLWriter

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Tested via remote streaming calls to dafilesrv with testsocket example that exposed bug.

### CODE REVIEW

File: rtl/eclrtl/rtlformat.cpp [hunk 1]
```diff
@@ -832,18 +832,12 @@ void CPropertyTreeWriter::outputQuoted(const char *text)
...
         return;
-    if (olen != len)
-    {
-        StringBuffer tmp(len, field);
-        target->setProp(fieldname, tmp.str());
-    }
-    else
-        target->setProp(fieldname, field);
+    StringBuffer tmp(len, field);
+    target->setProp(fieldname, tmp.str());
```
Comment (jakesmith) [inline rtl/eclrtl/rtlformat.cpp]: It is a bit of a shame to temporarily make a copy purely to ensure null terminated, for it then to be copied again inside the IPT call.
I did wonder whether I shouldn't introduce a IPropertyTree::setProp(const char *xpath, size32_t len, const char *val) .. and thereby avoid the need...

**Comment by jakesmith** [comment]
@ghalliday - please review and my comment

File: rtl/eclrtl/rtlformat.cpp (hunk 1 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtlformat.cpp]: Seems like a reasonable thing to add to me...

Comment (richardkchapman) [reply rtl/eclrtl/rtlformat.cpp]: (I suspect there are other places where it would have been used if available)

Comment (ghalliday) [reply rtl/eclrtl/rtlformat.cpp]: A new function would make sense, but this change looks ok.

**Pull request merged**